docs/caja-autorun-software.1
.\" Man page for caja-autorun-software
.\"
.\" This is free software; you may redistribute it and/or modify
.\" it under the terms of the GNU General Public License as
.\" published by the Free Software Foundation; either version 2,
.\" or (at your option) any later version.
.\"
.\" This is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\"You should have received a copy of the GNU General Public License along
.\"with this program; if not, write to the Free Software Foundation, Inc.,
.\"51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
.\"
.TH CAJA-AUTORUN-SOFTWARE 1 "29 January 2014" "MATE Desktop Environment" "General Manual"
.\" Please adjust this date whenever revising the manpage.
.\"
.SH "NAME"
caja-autorun-software \- Autorun Software in Caja
.SH "DESCRIPTION"
This executable is provided as a part of the Caja package for the MATE Desktop Environment.
.SH "BUGS"
.SS Should you encounter any bugs, they may be reported at: 
https://github.com/mate-desktop/caja/issues
.SH "LICENSE"
GPL v2 or later
.SH "SEE ALSO"
.SS
Caja documentation can be found from the "Help" menu, or by pressing the F1 key. 
Further information may also be available at: https://wiki.mate-desktop.org/docs
.P
.BR "caja" (1),
.BR "caja-connect-server" (1),
.BR "caja-file-management-properties" (1)


NEWS
### caja 1.23.0

  * update translations
  * caja-property-browser: Fix cast from non-struct type to struct type
  * eggsmclient-xsmp: Fix cast from non-struct type to struct type
  * caja-file-utilities: Fix cast from non-struct type to struct type
  * local #include files who are libraries inside <> instead ""
  * eel-string: Use 'va_copy' instead of 'G_VA_COPY'
  * eel-editable-label: Use 'memmove' instead of 'memcpy'
  * eel: remove eel_g_object_list_ref
  * drop functions defined but not used
  * local #include files inside "" instead <>
  * caja-icon-canvas-item: avoid 'cppcheck' warning
  * caja-file: avoid shifting signed 32-bit value by 31 bits
  * Fix invocation with --geometry from the command line
  * caja-icon-canvas-item: initialize variable 'scale'
  * Pass "NULL" for keyboard event
  * caja-application: avoid NULL inside 'g_object_unref'
  * fm-desktop-icon-view: remove dead code
  * caja-file-operations: make sure 'error' isn't NULL
  * caja-open-with-dialog: make sure there is error before showing it
  * eel-string: Fix: 'memcpy' overflows destination buffer
  * eel-string: avoid 'memcpy' with NULL argument
  * caja-progress-info: make sure argument is initialized
  * eel: replace eel_ref_str with GRefString
  * eel: remove g_object_list_copy
  * caja-sidebar-title: avoid 'NULL' inside 'setup_gc_with_fg'
  * caja-file: Fix incomplete file replacement dialog
  * caja-autorun: avoid NULL in 'memcpy'
  * eel-editable-label: call argument with initialized value
  * caja-file: Fix: true and false branches are identical
  * caja-file-operations: add comment with false positive warning:
  * [ci] use clang analyzer
  * configure.ac: fix portability issue with test
  * eel: remove g_str_list_copy
  * Remove USE_MATE2_MACROS from autogen.sh (legacy)
  * eel: remove g_list_free_deep_custom
  * avoid variables set but not used
  * eel-canvas: Fix warning: Undefined or garbage value returned to caller
  * caja-window: Fix garbage value
  * eel-gdk-pixbuf-extensions: Fix division by zero
  * caja-file: Fix use of memory after it is freed
  * caja-icon-container: Fix use of memory after it is freed
  * Fix attempt to free released memory
  * fm-tree-model: Fix use of memory after it is freed
  * caja-information-panel: Fix use of memory after it is freed
  * fm-directory-view: Fix use of memory after it is freed
  * caja-tree-view-drag-dest: Fix memory leak
  * [Security] eel-string: Use 'g_strlcat' instead of 'strcat'
  * [Security] Use 'g_strlcpy' instead of 'strcpy'
  * drop unused variables

### caja 1.22.0

  * update translations
  * Add update-authors.pl script which helps to update caja.about
  * Read authors and documenters fom caja.about file
  * Ensure proper translation of the about dialog title
  * Update authors
  * Update Documenters
  * Add --disable-self-check for configure

### caja 1.21.4

  * update translations
  * Avoid deprecated g_type_class_add_private under src
  * Fix incompatible pointer types
  * Avoid deprecated g_type_class_add_private under src
  * Avoid deprecated g_type_class_add_private under libcaja-private
  * Enable Travis CI
  * Fix compile warning
  * Let file size units follow the users preference
  * Fixed the value of 6 month and 1 year.
  * Support querying files by modification time and size
  * Avoid deprecated g_type_class_add_private
  * Update copyright to 2019
  * CajaApplication: avoid deprecated 'g_type_class_add_private'
  * caja-menu: avoid deprecated 'g_type_class_add_private'
  * [eel][libegg] Use G_DEFINE_TYPE_WITH_PRIVATE macro
  * configure.ac: Add check for PKG_CONFIG being set.
  * caja-application: Check for smclient before trying to load session
  * [eel] Avoid deprecated g_type_class_add_private
  * [libegg] Avoid deprecated g_type_class_add_private
  * [libcaja-private] Re-check file MIME type before picking an application.
  * [file-manager] Fix apparent bug in context menu updating code.
  * Revert "fm-directory-view.c:  Decrease minimum update interval."
  * Fix: weird item in bookmarks if the folder was removed
  * Minor code-quality enhancements
  * Forgot to initialize is_backup field
  * Change show/hide backup files shortcut
  * Remove useless comments, correct description string
  * Skip hidden files only, not backups
  * Add independent show/hide backup files
  * deleting unused nautilus icon
  * caja-progress-info: new behavior with notifications:
  * caja-progress-info: show notifications with the gsettings key:
  * move appdata to metainfo directory
  * disable deprecation warnings for distcheck
  * caja-places-sidebar: replace two other uses of g_drive_poll_for_media_finish
  * Fix segfault on stopping USB hard drives
  * caja-bookmark: Fix memory leak
  * fm-properties-window: Fix memory leaks
  * caja-desktop-link-monitor: Fix memory leak
  * eel-gtk-extensions: Fix memory leaks

### caja 1.21.3

  * GSettings should init beforce use it
  * update translations
  * Fix case-insensitive comparison in running_in_mate
  * monitor: fix crash on unmounting remote filesystems
  * CajaFile: fix crash in modify_link_hash_table
  * connect-server-dialog: Respect password save setting
  * properties-window: fix checking inconsistent state
  * monitor: watch for removal of non-native mounts on GVolumeMonitor
  * Add 'Configurable' interface and add support to caja-file-management-properties
  * caja-application.c: Remove redundant empty lines
  * Update caja manpage with info about "--force-desktop" option
  * Caja-icon-info: Optimize icon lookups by GIcon.
  * fm-directory-view.c:  Decrease minimum update interval.
  * Remove dead code, fix a memory leak
  * add option to open multiple uri in tabs at startup
  * pathbar: Activate pane when changing location
  * add the ability to switch tabs using [ctrl+tab] and [ctrl+shift+tab]
  * caja-window: Fix: mouse doesn't work with [ctrl] + [.] or [ctrl] + [;]
  * caja-desktop-directory: stop segfaults in directory_ready_callback
  * caja-directory.c: fix build warning

src/caja-autorun-software.c
/* -*- Mode: C; indent-tabs-mode: t; c-basic-offset: 8; tab-width: 8 -*- */

/* Caja

   Copyright (C) 2008 Red Hat, Inc.

   The Mate Library is free software; you can redistribute it and/or
   modify it under the terms of the GNU Library General Public License as
   published by the Free Software Foundation; either version 2 of the
   License, or (at your option) any later version.

   The Mate Library is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   Library General Public License for more details.

   You should have received a copy of the GNU Library General Public
   License along with the Mate Library; see the file COPYING.LIB.  If not,
   write to the Free Software Foundation, Inc., 51 Franklin St, Fifth Floor,
   Boston, MA 02110-1301, USA.

   Author: David Zeuthen <davidz@redhat.com>
*/


#include <config.h>
#include <unistd.h>
#include <string.h>
#include <time.h>
#include <errno.h>

#include <gtk/gtk.h>
#include <gio/gio.h>
#include <glib/gi18n.h>

#include "../libcaja-private/caja-module.h"
#include "../libcaja-private/caja-icon-info.h"

typedef struct
{
    GtkWidget *dialog;
    GMount *mount;
} AutorunSoftwareDialogData;

static void autorun_software_dialog_mount_unmounted (GMount *mount, AutorunSoftwareDialogData *data);

static void
autorun_software_dialog_destroy (AutorunSoftwareDialogData *data)
{
    g_signal_handlers_disconnect_by_func (G_OBJECT (data->mount),
                                          G_CALLBACK (autorun_software_dialog_mount_unmounted),
                                          data);

    gtk_widget_destroy (GTK_WIDGET (data->dialog));
    g_object_unref (data->mount);
    g_free (data);
}

static void
autorun_software_dialog_mount_unmounted (GMount *mount, AutorunSoftwareDialogData *data)
{
    autorun_software_dialog_destroy (data);
}

static gboolean
_check_file (GFile *mount_root, const char *file_path, gboolean must_be_executable)
{
    GFile *file;
    GFileInfo *file_info;
    gboolean ret;

    ret = FALSE;

    file = g_file_get_child (mount_root, file_path);
    file_info = g_file_query_info (file,
                                   G_FILE_ATTRIBUTE_ACCESS_CAN_EXECUTE,
                                   G_FILE_QUERY_INFO_NONE,
                                   NULL,
                                   NULL);
    if (file_info != NULL)
    {
        if (must_be_executable)
        {
            if (g_file_info_get_attribute_boolean (file_info, G_FILE_ATTRIBUTE_ACCESS_CAN_EXECUTE))
            {
                ret = TRUE;
            }
        }
        else
        {
            ret = TRUE;
        }
        g_object_unref (file_info);
    }
    g_object_unref (file);

    return ret;
}

static void
autorun (GMount *mount)
{
    char *error_string;
    GFile *root;
    GFile *program_to_spawn;
    char *path_to_spawn;
    char *cwd_for_program;

    root = g_mount_get_root (mount);

    /* Careful here, according to
     *
     *  https://standards.freedesktop.org/autostart-spec/autostart-spec-latest.html
     *
     * the ordering does matter.
     */

    program_to_spawn = NULL;
    path_to_spawn = NULL;

    if (_check_file (root, ".autorun", TRUE))
    {
        program_to_spawn = g_file_get_child (root, ".autorun");
    }
    else if (_check_file (root, "autorun", TRUE))
    {
        program_to_spawn = g_file_get_child (root, "autorun");
    }
    else if (_check_file (root, "autorun.sh", TRUE))
    {
        program_to_spawn = g_file_get_child (root, "autorun.sh");
    }
    else if (_check_file (root, "autorun.exe", TRUE))
    {
        /* TODO */
    }
    else if (_check_file (root, "AUTORUN.EXE", TRUE))
    {
        /* TODO */
    }
    else if (_check_file (root, "autorun.inf", FALSE))
    {
        /* TODO */
    }
    else if (_check_file (root, "AUTORUN.INF", FALSE))
    {
        /* TODO */
    }

    if (program_to_spawn != NULL)
    {
        path_to_spawn = g_file_get_path (program_to_spawn);
    }

    cwd_for_program = g_file_get_path (root);

    error_string = NULL;
    if (path_to_spawn != NULL && cwd_for_program != NULL)
    {
        if (chdir (cwd_for_program) == 0)
        {
            execl (path_to_spawn, path_to_spawn, NULL);
            error_string = g_strdup_printf (_("Error starting autorun program: %s"), strerror (errno));
            goto out;
        }
        error_string = g_strdup_printf (_("Error starting autorun program: %s"), strerror (errno));
        goto out;
    }
    error_string = g_strdup_printf (_("Cannot find the autorun program"));

out:
    if (program_to_spawn != NULL)
    {
        g_object_unref (program_to_spawn);
    }
    if (root != NULL)
    {
        g_object_unref (root);
    }
    g_free (path_to_spawn);
    g_free (cwd_for_program);

    if (error_string != NULL)
    {
        GtkWidget *dialog;
        dialog = gtk_message_dialog_new_with_markup (NULL, /* TODO: parent window? */
                 0,
                 GTK_MESSAGE_ERROR,
                 GTK_BUTTONS_OK,
                 _("<big><b>Error autorunning software</b></big>"));
        gtk_message_dialog_format_secondary_text (GTK_MESSAGE_DIALOG (dialog), "%s", error_string);
        gtk_dialog_run (GTK_DIALOG (dialog));
        gtk_widget_destroy (dialog);
        g_free (error_string);
    }
}

static void
present_autorun_for_software_dialog (GMount *mount)
{
    GIcon *icon;
    int icon_size;
    CajaIconInfo *icon_info;
    GdkPixbuf *pixbuf;
    GtkWidget *image;
    char *mount_name;
    GtkWidget *dialog;
    AutorunSoftwareDialogData *data;

    mount_name = g_mount_get_name (mount);

    dialog = gtk_message_dialog_new_with_markup (NULL, /* TODO: parent window? */
             0,
             GTK_MESSAGE_OTHER,
             GTK_BUTTONS_CANCEL,
             _("<big><b>This medium contains software intended to be automatically started. Would you like to run it?</b></big>"));
    gtk_message_dialog_format_secondary_text (GTK_MESSAGE_DIALOG (dialog),
            _("The software will run directly from the medium \"%s\". "
              "You should never run software that you don't trust.\n"
              "\n"
              "If in doubt, press Cancel."),
            mount_name);

    /* TODO: in a star trek future add support for verifying
     * software on media (e.g. if it has a certificate, check it
     * etc.)
     */


    icon = g_mount_get_icon (mount);
    icon_size = caja_get_icon_size_for_stock_size (GTK_ICON_SIZE_DIALOG);
    icon_info = caja_icon_info_lookup (icon, icon_size,
                                       gtk_widget_get_scale_factor (GTK_WIDGET (dialog)));
    pixbuf = caja_icon_info_get_pixbuf_at_size (icon_info, icon_size);
    image = gtk_image_new_from_pixbuf (pixbuf);
    gtk_widget_set_halign (image, GTK_ALIGN_CENTER);
    gtk_widget_set_valign (image, GTK_ALIGN_START);
    gtk_message_dialog_set_image (GTK_MESSAGE_DIALOG (dialog), image);

    gtk_window_set_title (GTK_WINDOW (dialog), mount_name);
    gtk_window_set_icon (GTK_WINDOW (dialog), pixbuf);

    data = g_new0 (AutorunSoftwareDialogData, 1);
    data->dialog = dialog;
    data->mount = g_object_ref (mount);

    g_signal_connect (G_OBJECT (mount),
                      "unmounted",
                      G_CALLBACK (autorun_software_dialog_mount_unmounted),
                      data);

    gtk_dialog_add_button (GTK_DIALOG (dialog),
                           _("_Run"),
                           GTK_RESPONSE_OK);

    gtk_widget_show_all (dialog);

    if (gtk_dialog_run (GTK_DIALOG (dialog)) == GTK_RESPONSE_OK)
    {
        gtk_widget_destroy (dialog);
        autorun (mount);
    }

    g_object_unref (icon_info);
    g_object_unref (pixbuf);
    g_free (mount_name);
}

int
main (int argc, char *argv[])
{
    GVolumeMonitor *monitor;
    GFile *file;
    GMount *mount;

    bindtextdomain (GETTEXT_PACKAGE, MATELOCALEDIR);
    bind_textdomain_codeset (GETTEXT_PACKAGE, "UTF-8");
    textdomain (GETTEXT_PACKAGE);

    gtk_init (&argc, &argv);

    if (argc != 2)
    {
        goto out;
    }

    /* instantiate monitor so we get the "unmounted" signal properly */
    monitor = g_volume_monitor_get ();
    if (monitor == NULL)
    {
        goto out;
    }

    file = g_file_new_for_commandline_arg (argv[1]);
    if (file == NULL)
    {
        g_object_unref (monitor);
        goto out;
    }

    mount = g_file_find_enclosing_mount (file, NULL, NULL);
    if (mount == NULL)
    {
        g_object_unref (file);
        g_object_unref (monitor);
        goto out;
    }

    present_autorun_for_software_dialog (mount);
    g_object_unref (file);
    g_object_unref (monitor);
    g_object_unref (mount);

out:
    return 0;
}
